src/CodeStyle/Core/CodeFixes/DocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,129 @@
+                        nameof(DocumentBasedFixAllProvider));
+                    break;
+
+                case FixAllScope.Custom:
+                default:
+                    fixAction = null;
+                    break;
+            }
+
+            return Task.FromResult(fixAction);
+        }
+
+        /// <summary>
+        /// Fixes all occurrences of a diagnostic in a specific document.
+        /// </summary>
+        /// <param name="fixAllContext">The context for the Fix All operation.</param>
+        /// <param name="document">The document to fix.</param>
+        /// <param name="diagnostics">The diagnostics to fix in the document.</param>
+        /// <returns>
+        /// <para>The new <see cref="SyntaxNode"/> representing the root of the fixed document.</para>
+        /// <para>-or-</para>
+        /// <para><see langword="null"/>, if no changes were made to the document.</para>
+        /// </returns>
+        protected abstract Task<SyntaxNode> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document, ImmutableArray<Diagnostic> diagnostics);
+
+        private async Task<Document> GetDocumentFixesAsync(FixAllContext fixAllContext)
+        {
+            var documentDiagnosticsToFix = await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(fixAllContext).ConfigureAwait(false);
+            if (!documentDiagnosticsToFix.TryGetValue(fixAllContext.Document, out var diagnostics))
+            {
+                return fixAllContext.Document;
+            }
+
+            var newRoot = await FixAllInDocumentAsync(fixAllContext, fixAllContext.Document, diagnostics).ConfigureAwait(false);
+            if (newRoot == null)
+            {
+                return fixAllContext.Document;
+            }
+
+            return fixAllContext.Document.WithSyntaxRoot(newRoot);
+        }
+
+        private async Task<Solution> GetSolutionFixesAsync(FixAllContext fixAllContext, ImmutableArray<Document> documents)
+        {
+            var documentDiagnosticsToFix = await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(fixAllContext).ConfigureAwait(false);
+
+            var solution = fixAllContext.Solution;
+            var newDocuments = new List<Task<SyntaxNode>>(documents.Length);
+            foreach (var document in documents)
+            {
+                if (!documentDiagnosticsToFix.TryGetValue(document, out var diagnostics))
+                {
+                    newDocuments.Add(document.GetSyntaxRootAsync(fixAllContext.CancellationToken));
+                    continue;
+                }
+
+                newDocuments.Add(FixAllInDocumentAsync(fixAllContext, document, diagnostics));
+            }
+
+            for (var i = 0; i < documents.Length; i++)
+            {
+                var newDocumentRoot = await newDocuments[i].ConfigureAwait(false);
+                if (newDocumentRoot == null)
```

Comment:
when does that happen?

Reply:
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/commit/ceb765cc63a4fb3669466eaf46b98e56d9ecea9f